lib/Sema/TypeCheckDeclOverride.cpp

```diff
@@ -928,6 +931,96 @@ static void checkOverrideAccessControl(ValueDecl *baseDecl, ValueDecl *decl,
   }
 }
 
+static GenericSignature *getNewGenericSignature(ValueDecl *base,
+                                                ValueDecl *derived) {
+  auto baseGenericCtx = base->getAsGenericContext();
+  auto derivedGenericCtx = derived->getAsGenericContext();
+  auto &ctx = base->getASTContext();
+
+  if (!baseGenericCtx) {
+    return nullptr;
+  }
+
+  auto baseClass =
+      baseGenericCtx->getInnermostTypeContext()->getSelfClassDecl();
+  auto derivedClass =
+      derivedGenericCtx->getInnermostTypeContext()->getSelfClassDecl();
+
+  if (!derivedClass) {
+    return nullptr;
+  }
+
+  if (!derivedClass->getInterfaceType()->is<LValueType>()) {
+    return nullptr;
+  }
+
+  auto subMap = derivedClass->getInterfaceType()->getContextSubstitutionMap(
+      derivedClass->getModuleContext(), baseClass);
+
+  auto *genericParams = baseGenericCtx->getGenericParams();
+  if (genericParams) {
+    SmallVector<GenericTypeParamDecl *, 4> newParams;
+
+    // First, clone the superclass constructor's generic parameter list,
+    // but change the depth of the generic parameters to be one greater
+    // than the depth of the subclass.
+    unsigned depth = 0;
+    if (auto *genericSig = baseClass->getGenericSignature())
+      depth = genericSig->getGenericParams().back()->getDepth() + 1;
+
+    for (auto *param : genericParams->getParams()) {
+      auto *newParam = new (ctx) GenericTypeParamDecl(
+          baseClass, param->getName(), SourceLoc(), depth, param->getIndex());
+      newParams.push_back(newParam);
+    }
+
+    GenericSignatureBuilder builder(ctx);
+    builder.addGenericSignature(derivedClass->getGenericSignature());
+
+    auto source =
+        GenericSignatureBuilder::FloatingRequirementSource::forAbstract();
+    auto *baseClassSig = baseClass->getGenericSignature();
+
+    unsigned superclassDepth = 0;
+    if (baseClassSig)
+      superclassDepth =
+          baseGenericCtx->getGenericParams()->getParams().back()->getDepth() +
+          1;
+
+    // We're going to be substituting the requirements of the base class
+    // initializer to form the requirements of the derived class initializer.
+    auto substFn = [&](SubstitutableType *type) -> Type {
+      auto *gp = cast<GenericTypeParamType>(type);
+      if (gp->getDepth() < superclassDepth)
+        return Type(gp).subst(subMap);
+      return CanGenericTypeParamType::get(
+          gp->getDepth() - superclassDepth + depth, gp->getIndex(), ctx);
+    };
+
+    auto lookupConformanceFn =
+        [&](CanType depTy, Type substTy,
+            ProtocolDecl *proto) -> Optional<ProtocolConformanceRef> {
+      if (auto conf = subMap.lookupConformance(depTy, proto))
+        return conf;
+
+      return ProtocolConformanceRef(proto);
+    };
+
+    for (auto reqt : baseClassSig->getRequirements())
```

Comment:
This should be looping over the base method's requirements, not the base class requirements, so eg baseGenericCtx->getGenericSignature()->getRequirements()